browser/ipfs/ipfs_tab_helper.cc

```diff
@@ -62,19 +105,96 @@ bool IPFSTabHelper::MaybeCreateForWebContents(
   return true;
 }
 
+void IPFSTabHelper::HostResolvedCallback(
+    const std::string& host,
+    const std::vector<std::string>& text_results) {
+  std::string dnslink = GetDSNRecordValue(text_results, kDnsLinkHeader);
+  if (dnslink.empty())
+    return;
+  GURL current = web_contents()->GetURL();
+  if (current.host() != host)
+    return;
+  ipfs_resolved_host_ = host;
+}
+
+GURL IPFSTabHelper::GetIPFSResolvedURL() const {
+  if (ipfs_resolved_host_.empty())
+    return GURL();
+  GURL current = web_contents()->GetURL();
+  GURL::Replacements replacements;
+  replacements.SetSchemeStr(kIPNSScheme);
+  return current.ReplaceComponents(replacements);
+}
+
+bool IPFSTabHelper::HasActiveResolverForHost(const std::string& host) const {
+  for (auto& resolver : resolvers_) {
+    if (!resolver->IsRunning())
+      continue;
+    if (resolver->host() == host)
+      return true;
+  }
+  return false;
+}
+
+void IPFSTabHelper::ResolveIPFSLink() {
+  if (web_contents()->GetBrowserContext()->IsTor() ||
+      web_contents()->GetBrowserContext()->IsOffTheRecord())
+    return;
+
+  GURL current = web_contents()->GetURL();
+  if (ipfs_resolved_host_ == current.host())
+    return;
+
+  if (HasActiveResolverForHost(current.host()))
+    return;
+
+  auto* storage_partition = content::BrowserContext::GetDefaultStoragePartition(
+      web_contents()->GetBrowserContext());
+
+  const auto& host_port_pair = net::HostPortPair::FromURL(current);
+  for (auto& resolver : resolvers_) {
+    if (resolver->IsRunning())
+      resolver->Stop();
+
+    auto resolved_callback = base::BindOnce(
+        &IPFSTabHelper::HostResolvedCallback, weak_ptr_factory_.GetWeakPtr());
+
+    resolver->Resolve(host_port_pair, net::NetworkIsolationKey(),
+                      storage_partition->GetNetworkContext(),
+                      net::DnsQueryType::TXT, std::move(resolved_callback));
+  }
+}
+
 void IPFSTabHelper::DidFinishNavigation(content::NavigationHandle* handle) {
   DCHECK(handle);
-  if (!handle->IsInMainFrame()) {
+  if (!handle->IsInMainFrame() || !handle->HasCommitted() ||
+      !handle->GetResponseHeaders()) {
+    return;
+  }
+
+  if (!handle->IsSameDocument()) {
+    ipfs_resolved_host_.erase();
+  }
+
+  int response_code = handle->GetResponseHeaders()->response_code();
+  if (ipfs_resolved_host_.empty() &&
+      response_code >= net::HttpStatusCode::HTTP_INTERNAL_SERVER_ERROR &&
+      response_code <= net::HttpStatusCode::HTTP_GATEWAY_TIMEOUT) {
+    ResolveIPFSLink();
```

Comment:
why is this called in this case?

Reply:
https://github.com/brave/brave-browser/issues/13609#issuecomment-764627805
```
or if page returned code different from OK 200 or failed due to network/DNS (eg. invalid A record)
(this will make it possible for people to recover from broken/DDoSs HTTP servers etc)
```